### ISSUE TITLE
About prototype_separation_loss

### ISSUE THREAD

**Issue by Yisikaliou** [mashijie1028/ProtoGCD#4]
Thank you for your excellent work. When I reproduced it, I found that the prototype_separation_loss was negative value. What's the reason for this? Is this normal?

**Comment by mashijie1028** [comment]
Yes, it is normal for the prototype_separation_loss $L_{sep}$ to be negative. Loss functions are not required to be positive in deep learning. In our case, $L_{sep}=\frac{1}{K}\sum_i\log\frac{1}{K-1}\sum_{j\neq i}\exp(\mu_i^\top\mu_j)/\tau$ (see Eq. (14) in our paper). As training proceeds, the similarity between prototypes $\mu_i$ and $\mu_j$ becomes lower. When $\mu_i^\top\mu_j<0$, $\exp(\cdot)<1$ and finally $L_{sep}=\log\sum\exp(\cdot)<0$.

**Comment by Yisikaliou** [comment]
Thank you very much for your reply. Your answer has dispelled my doubts. In addition, what I want to confirm is that the performance reported in the paper is based on the best-performing checkpoints in the training set or the best checkpoints selected based on the test set. Or perhaps it is the performance of the training set (test set) from the last epoch

**Comment by mashijie1028** [comment]
We mainly report the performance of the last epoch. Note that GCD is a transductive setting, where the evaluation is conducted on the unlabeled training samples.

**Comment by Yisikaliou** [comment]
I see. Thank you for your response！
